### ISSUE TITLE
Command reorganization/discovery/reusability

### ISSUE THREAD

**Issue by Xcelled** [aura-project/aura#77]
Since there are so many commands, it might be helpful to group them into namespaces. For example, `char.height` and `item.drop`. We could map frequently used commands via aliases, like `item` => `item.spawn`. One could even go to extremes like `char.position.region.blah.lasttown`.

Thus could also allow us to specialize commands. So we could have a `go.uladh` and a `go.iria`, for example.

**Comment by exectails** [comment]
I'm against this. Typing commands, especially repeatedly, is annoying already, to the extend that every time I have to type something repeatedly, I'm thinking about adding an alias for it. Why would you choose to go with `item.spawn`, if `item` sufficient, more intuitive, and shorter?

**Comment by exectails** [comment]
Does anybody else have an opinion about this?

**Comment by DeleteIT** [comment]
The *only* way I'd personally be on board is if you could treat commands like Intellisense treats namespaces, for example, sending "item" and getting back all the possible follow on namespaces, and continuing my progress in this fashion until I reach the command of my choice.
Which would obviously be too unnecessarily complex for the server to keep track of.

But then, it would be more useful to just have a >help function, that lists all the commands, each command to a line, perhaps with a *short* description of what they do, but pushed to the Trade chat tab, with a System message telling them to move to that tab.

Or having everything grouped into categories, and `>help` returning those categories, whilst `>help  move` returning as per above, but with all commands pertaining to moving things around.

Ultimately, what is being suggested can be better achieved with switches and alias commands `>item /id:2000 /n:20 /p:floor` becoming `>itemdrop /id:2000 /n:20` instead of >item.drop /id:2000 /n:20`, since I only need to know one command to *function*, but if I want to be more efficient, I can break out alias commands.
Additionally, I'd suggest `>alias [command]` to facilitate users learning what commands have aliases.
In the end, I think that any end user who has never opened VS will be hopelessly confused, and subsequently frustrated.

Basically, it seems like a good idea, until you realise that you'd really have to find commands like:
`>char {enter}`
`>char.position {enter}`
`>char.position.region {enter}`
`>char.position.region.blah {enter}`
`>char.position.region.blah.lasttown {enter}`

Also, isn't `go.Iria` just `>go Iria` with a different separator?

**Comment by DeleteIT** [comment]
Also, on the subject of continually repeating myself with commands, `>repeat` would be pretty awesome, for clients that don't have basic edit/history functionality (looking at you, NA client, the *ONLY* client missing that functionality..).

**Comment by Cosrnos** [comment]
I think the command structure is fine as is.

We have to make the assumption that most people using these commands is not familiar with code and the namespace structure is unnecessary bloat. Commands should be easy to quickly issue and as short as possible IMHO.

If we're concerned about too many commands and needing a better way to organize them, I think the answer would be in a custom GM Panel UI or something of the sort.

**Comment by DeleteIT** [comment]
I don't think we can even add to the UI with any real ease, if you *did* mean a client based one.

**Comment by Cosrnos** [comment]
I did not mean a client based one :P

It's not a fully formed idea, but the concept of having an additional client open that connects to the server and requires you to log in again. The server could validate the requests are coming from the same IP and the character is already logged in before connecting the admin panel.

Again, just an aside and I don't think it would be wise to pursue anytime soon, but it's a thing we *could* do.

**Comment by DeleteIT** [comment]
I thought you might mean something like that, and I think that's a much more workable solution to admin woes than anything else, especially since it could offer a much greater level of information than the client does currently, without even having the GM panel open.

**Comment by exectails** [comment]
>repeat would be pretty awesome

Yea, I've been thinking about that for a while, didn't mention it here because repeat isn't really related to namespaces. Every time I have to repeat a command I think "hm... >rpt", and the rest of the time I'm thinking "nah, that's not worth it".

>additional application

That's an interesting idea, although, I can't think of too many functions that we currently don't have, that such an application might handle. It would basically be another GMCP, what would be in there, that our commands and the official CP don't handle yet?

**Comment by DeleteIT** [comment]
A list of all currently online players, grouped by their region. Or level. Or auth level.
The ability to click that player name, and get information, where they are, what their stats are, what they're wearing, recent chat/command history, what they are currently targeting, whether they're in a party, etc.
You could then interact with them, with a click, for example, if I wanted to send a party to a dungeon instance that doesn't exist yet, I could achieve that with a handful of clicks, based around interacting with the party, instead of a slew of typed commands that will probably involve interacting with each player singularly, and be moderately complex by comparison.

It would basically be an informative alias system.

**Comment by Cosrnos** [comment]
The initial goal for the app wouldn't be to introduce any new functionality, but rather to organize the commands in an intuitive way. Eventually what he said ^

**Comment by Xcelled** [comment]
For what it's worth, I played with a new command manager here: https://github.com/Xcelled/aurora/blob/master/Aurora/Commands/AuroraCommandManager.cs. It supports intellisense-style completion, the help command, and command lists. Of course, there are no actual commands beyond these, but whatever.

I was considering implementing a form like @Cosrnos said, but I rejected it because it would either be limited to the local machine running the server (limiting usefulness), or require a whole new RMI-style subsystem. (SMI? Server Method Invocation? Hehehe). But such a system could probably be cobbled together from one of the remoting packages if we don't want to write a whole API.

One thing @DeleteIT has touched on but not explicitly stated is parameters. Right now, commands are all based on position only. Maybe we should explore unix-like parameters? For example, to specify color 3 in the item command, you also have to specify 1 and 2, instead of something like `>item foo -c3:ffffff`.

**Comment by DeleteIT** [comment]
I didn't specifically state it, because I forgot it wasn't a thing in Aura.
That should absolutely become a thing, since it slashes the need for a lot of aliases.

**Comment by exectails** [comment]
>it slashes the need for a lot of aliases.

Not really, the aliases we have are just for convenience, except for `drop`. It would make extension easier though, when a command requires new parameters. I've thought about that before, but the change would require some work, and for *most* commands it's not really required at all. The only commands I can think of atm, that would really benefit are `item` and `spawn`.

>organize the commands in an intuitive way

I really don't see anyone using a second "client", just to, for example, spawn an item more "easily". Especially if that means that you *have to* run your client in window mode, and constantly switch between windows.

There is learning involved with commands of course. A newbie does not go into a server, knowing how the commands work, they have to learn them and their parameters. That's why advanced newbies are commonly proud of knowing all the commands, as if that would make them good GMs or something. Those people could maybe benefit from a GUI, where they can type/select exactly what they want to happen, but is that worth the trouble?

Personally I wouldn't want to devote too much time to such a project, and I don't think it should be part of Aura's core, as it's just a helper tool for a select few.

**Comment by DeleteIT** [comment]
Scenario:

I want to put a party into a new dungeon.


Typed commands:
Currently, the functionality to create a dungeon manually doesn't exist, but if it did, it would probably look like `>createDungeon [seed] [comma separated list of names/IDs of people who made the dungeon] | [partyID of the party who made the dungeon] [dungeon] [level]`
Which also brings up the point that I need the partyID, because I can't refer to this party by the leader, since that could change moment to moment.
So, I need another currently non-existent command, `>getPartyID [playerName]`, and then to remember that ID, before using that to create the dungeon.

Then I'd have to move the party.
So, I either write a `>moveParty [partyID]` function, to shift all members at the same time, but in this particular case, I don't want to move one of the members.
So I have to move them all singularly.
We'll assume that the dungeon creation method gave us that region ID.

GM GUI:
Click player.
Click party.
Click `move`.
Select destination as `instance`.
Click `new`.
Select params from contextual drop downs.
Select members to move to instance.
Click `done` to action your command.


I could have done that from my phone, via a Unity based application, for example.

**Comment by DeleteIT** [comment]
Your item example, by the way?
Colour picker to pick out my colour selection.
Enchant selection.
Reforge selection.
Upgrade selection.
Stat edit.

And so on.

**Comment by Cosrnos** [comment]
Even having a web interface that the server spawns on a specific port is an option, with a mobile friendly interface :+1:

**Comment by DeleteIT** [comment]
And don't even get me *started* on doing anything in bulk, with the NA client's inability to repeat an effing command with the simple expedient of pressing the up key.

**Comment by exectails** [comment]
>the functionality to create a dungeon manually doesn't exist

It does, the command is called `dungeon`. Even if it doesn't support all of what you're asking for.

Since the scenario you're giving here is so specific, I'll treat it as an actual thing someone might want to do for a moment. My comments are:

* The way you're imagining the dungeon command is wrong, there would be no level, there probably would be no need for a seed, and there probably wouldn't be a need for a list of creators.
* You would not need the party id, because you could get the party by naming any one member of the party, which would also give you the creators.
* For adding or removing people, you could create a kind of variable arguments for this command, prefixing the names with - to exclude them or something.
* You would not need a move command, because that could happen automatically, when creating the dungeon.
* So in reality you would probably only need a single command, `dungeon`. All it would need is one simple change, it would need to warp the entire party, and not only the person you ran the command on: `>>dungeon Character tircho_alby_dungeon`

You probably only meant that as an example, but it doesn't really hold up, because it can be much simpler than you described (see above), and the example is very, *very* specific. It's not something anyone would even actually do I think, I mean, why would you? It might be something you would want to do during an event, but then you wouldn't use a command for it.

Of course, I do acknowledge that, *if* something was as complicated as your example, and *if* you needed to be this specific, selecting only specific people of a specific party, the command system wouldn't be sufficient. But that wouldn't *be* a job for the command system then, as it's way more than a simple *command* to the server, it's a series of tasks. And if you have a need for something so specific, you would have to integrate it into the application and the server anyway, which means you're capable enough to do it without the tool existing first, right now, via any number of ways, like, for example, a web interface, that could even be spawned by a script.

This is a nice idea, most certainly, just not something that belongs to the *core* in my opinion, don't you agree?

>Your item example, by the way?

Okay, some people might find that useful. Personally I'm someone who wouldn't even grant my GMs an item command, *never* ends well, and I don't see a need for them to be able to spawn items with specific enchants, reforges, or upgrades either. *If* you do have the need to constantly spawn such specifically crafted items, then such a tool would indeed be great.

**Comment by DeleteIT** [comment]
/Shrug

>Especially if that means that you have to run your client in window mode, and constantly switch between windows.

Find me a GM, or even just a long time player who even plays the game in full screen.

**Comment by exectails** [comment]
>Find me a GM, or even just a long time player who even plays the game in full screen.

Just saying, it is a restriction.

**Comment by Cosrnos** [comment]
The point isn't whether *you* would grant a GM access to these commands or whether *you* find the commands easy enough, it's about providing end users, people who possibly have 0 technical experience with an experience that is simple enough for them to use.

I'm not arguing this project belongs in the core--on the contrary. I'm saying the core handles item commands smoothly as far as we need to be concerned, but there's a gap that could easily be filled by another tool.

**Comment by exectails** [comment]
>I'm not arguing this project belongs in the core--on the contrary. I'm saying the core handles item commands smoothly as far as we need to be concerned, but there's a gap that could easily be filled by another tool.

That's exactly the point I was trying to make, and that does play into my reasoning. Of course it's about the end-user in the end, but I do have to be a little selective about what I work on. Our resources are very limited, and I don't want to "waste" time on adding something to the core that doesn't belong there, or isn't really necessary for most people.

Like I said, such an application *can* be great, *if* you have the need for it. But it's not something I will work on.

**Comment by Cosrnos** [comment]
Then I don't know what we're arguing about :)

**Comment by exectails** [comment]
Didn't realize we were arguing =)

The other thing though, named parameters for the commands, is a thing we would probably want. item and spawn are a little limited atm, and there are actually even things you can't do with item currently, like spawning a gold pouch with a specific amount of gold in it.

**Comment by DeleteIT** [comment]
I don't think anyone was suggesting that you should work on it personally, @exectails, in fact, I was going to say that the ***absolute*** most you should do is add some kind of authentication for it, and *maybe* write a demo/base app with ONE button, which is "Hello World!", which sends same.

I'm sure the many people who know Winforms/Unity/whatever it's made in will expand on it, as demand for such functionality grows.

As for categorising, I'd just generalise, `item`, `move`, `create`, really distil down to the most basic essence of what it is/does.
You can always have things in multiple categories, and if people don't agree with the categories default assigned, they could always write their own, if you load the category information from the options.

**Comment by exectails** [comment]
>I don't think anyone was suggesting that you should work on it personally

You'll understand that I argue with the thought of *me* implementing it when anything is suggested here^^

The web API was basically supposed to do stuff like this, it just has to be taught to do them. But it will never be as easy as just adding to the application, the server has to support it as well, forward the commands to the correct channels, etc.

**Comment by DeleteIT** [comment]
With things like this, that aren't super core related, I'd expect you to do pretty much what I said, put in the most basic implementation and then drop it on the public.

**Comment by ImaBrokeDude** [comment]
This reminds me of what I wanted to do with InstaConfig, it would basically just be a supplemental thing though we would at least leave a standard in which any tool written by anyone would follow... the accessibility should be core but the actual implementation or use can be seperate... hopefully I made sense.

**Comment by exectails** [comment]
I'll close this, as it became about a tool that will have nothing to do with the core, if anybody decides to create it. Instead I'll reopen the `>help` suggestion.
